scrapy/pqueues.py

```diff
@@ -0,0 +1,217 @@
+        self.pqueues = {}  # slot -> priority queue
+        for slot, startprios in (slot_startprios or {}).items():
+            self.pqueues[slot] = self.pqfactory(startprios)
+
+    def pop_slot(self, slot):
+        """ Pop an object from a priority queue for this slot """
+        queue = self.pqueues[slot]
+        request = queue.pop()
+        if len(queue) == 0:
+            del self.pqueues[slot]
+        return request
+
+    def push_slot(self, slot, obj, priority):
+        """ Push an object to a priority queue for this slot """
+        if slot not in self.pqueues:
+            self.pqueues[slot] = self.pqfactory()
+        queue = self.pqueues[slot]
+        queue.push(obj, priority)
+
+    def close(self):
+        active = {slot: queue.close()
+                  for slot, queue in self.pqueues.items()}
+        self.pqueues.clear()
+        return active
+
+    def __len__(self):
+        return sum(len(x) for x in self.pqueues.values()) if self.pqueues else 0
+
+    def __contains__(self, slot):
+        return slot in self.pqueues
+
+
+class DownloaderAwarePriorityQueue(object):
+
+    _DOWNLOADER_AWARE_PQ_ID = 'DOWNLOADER_AWARE_PQ_ID'
+
+    @classmethod
+    def from_crawler(cls, crawler, qfactory, startprios=None):
+        return cls(crawler, qfactory, startprios)
+
+    def __init__(self, crawler, qfactory, startprios=None):
+        ip_concurrency_key = 'CONCURRENT_REQUESTS_PER_IP'
+        ip_concurrency = crawler.settings.getint(ip_concurrency_key, 0)
+
+        if ip_concurrency > 0:
+            raise ValueError('"%s" does not support %s=%d' % (self.__class__,
+                                                              ip_concurrency_key,
+                                                              ip_concurrency))
+
+        def pqfactory(startprios=()):
+            return PriorityAsTupleQueue(qfactory, startprios)
+
+        if startprios and not isinstance(startprios, dict):
+            raise ValueError("DownloaderAwarePriorityQueue accepts "
+                             "``startprios`` as a dict; %r instance is passed."
+                             " Only a crawl started with the same priority "
+                             "queue class can be resumed." % startprios.__class__)
+        self._slot_pqueues = SlotPriorityQueues(pqfactory,
+                                                slot_startprios=startprios)
+
+        self._active_downloads = {slot: 0 for slot in self._slot_pqueues.pqueues}
+        crawler.signals.connect(self.on_response_download,
+                                signal=response_downloaded)
+        crawler.signals.connect(self.on_request_reached_downloader,
+                                signal=request_reached_downloader)
+
+    def mark(self, request):
+        meta = _get_request_meta(request)
+        if not isinstance(meta, dict):
+            raise ValueError('No meta attribute in %s' % (request, ))
+        meta[self._DOWNLOADER_AWARE_PQ_ID] = id(self)
```

Comment:
please also remove this meta key properly, for instance in "on_response_download"